examples/optimizers/Main.hs

```diff
@@ -0,0 +1,226 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Main where
+
+{- Optimizer implementations run on standard test functions -}
+
+import Control.Monad (foldM, when)
+import GHC.Generics
+import Prelude hiding (sqrt)
+import Text.Printf (printf)
+
+import Torch.Tensor
+import Torch.TensorFactories (eye', zeros')
+import Torch.Functions
+import Torch.Autograd
+import Torch.NN
+
+import TestFunctions
+
+-- Optimizers
+
+type LearningRate = Tensor
+type Gradient = [Tensor]
+
+-- gradient descent step
+gd :: LearningRate -> [Parameter] -> Gradient -> [Tensor]
+gd lr parameters gradients = zipWith step depParameters gradients
+  where
+    step p dp = p - (lr * dp)
+    depParameters = fmap toDependent parameters
+
+-- gradient descent with momentum step
+--     lr              beta      memory      parameters     gradients
+gdm :: LearningRate -> Float -> [Tensor] -> [Parameter] -> Gradient -> [(Tensor, Tensor)]
+gdm lr beta gradMemory parameters gradients = (zipWith3 step) depParameters gradients gradMemory
+  where
+    z' dp z = cmul z beta + dp
+    step p dp z = let newZ = z' dp z in (p - lr * newZ, newZ)
+    depParameters = fmap toDependent parameters
+
+-- Adam
+
+-- | State representation for Adam Optimizer
+data Adam = Adam { 
+    m1 :: [Tensor], -- 1st moment
+    m2 :: [Tensor], -- 2nd moment
+    modelParam :: [Tensor] 
+    } deriving Show
+
+-- | Adap step
+adam :: LearningRate -> Float -> Float -> Adam -> Gradient -> Int -> Adam
+adam lr beta1 beta2 Adam{..} gradients iter = Adam m1' m2' w
+    where
+        -- 1st & 2nd moments
+        f1 m1 dp = cmul m1 beta1 + cmul dp (1 - beta1)
+        f2 m2 dp = cmul m2 beta2 + cmul (dp * dp) (1 - beta2)
+        m1' = zipWith f1 m1 gradients
+        m2' = zipWith f2 m2 gradients
+        -- averages of moments
+        a beta m = cdiv m (1 - beta^(iter + 1))
+        a1 = fmap (a beta1) m1'
+        a2 = fmap (a beta2) m2'
+        -- parameter update
+        eps = 1e-15
+        fw wprev avg1 avg2 = wprev - lr * avg1 / (sqrt avg2 + eps)
+        w = zipWith3 fw modelParam a1 a2
+
+-- | show output after n iterations
+showLog :: (Show a) => Int -> Int -> Tensor -> a -> IO ()
+showLog n i lossValue state = 
+    when (mod i n == 0) do
+        putStrLn ("Iter: " ++ printf "%4d" i 
+            ++ " | Loss:" ++ printf "%.4f" (asValue lossValue :: Float)
+            ++ " | Parameters: " ++ show state)
+
+-- | produce flattened parameters and gradient for a single iteration
+runIter state loss i = do
+    showLog 1000 i lossValue state
+    pure (flattenParameters state, grad lossValue flatParameters)
+    where
+        lossValue = loss state
+        flatParameters = flattenParameters state
+        gradients = grad lossValue flatParameters
+
+-- | foldM as a loop with action block as the last argument
+foldLoop :: Monad m => a -> Int -> (a -> Int -> m a) -> m a
+foldLoop x count block = foldM block x [1..count]
+
+-- | Rosenbrock function gradient descent
+testRosenGD :: Int -> IO ()
```

Comment:
could this be translated into a unit test?
this doesn't seem to be deterministic:
```
*TestFunctions Main> testRosenGD 10
Initial :(0.32480815,1.086431)
*TestFunctions Main> testRosenGD 10
Initial :(0.6006327,0.91784304)
```

Reply:
Definitely can. The non-determinism comes from the initialization being random. Would be fairly straightforward to incorporate these into a unit test (though we might hold off on extensive testing of this sort for the library version of the API to take shape).

Reply:
It might be also interested to explore some [avenues that JAX took](https://github.com/google/jax/blob/master/design_notes/prng.md). Their approach to non-determinism is heavily functional.